botocore/configprovider.py

```diff
@@ -851,3 +858,149 @@ def provide(self):
+    changes or removal without prior announcement. Please do not use it
+    directly.
+    """
+
+    _ENDPOINT_URL_LOOKUP_ORDER = [
+        'environment_service',
+        'environment_global',
+        'config_service',
+        'config_global',
+    ]
+
+    def __init__(
+        self,
+        full_config,
+        scoped_config,
+        client_name,
+        environ=None,
+    ):
+        """Initialize a ConfiguredEndpointProviderChain.
+
+        :type full_config: dict
+        :param full_config: This is the dict representing the full
+            configuration file.
+
+        :type scoped_config: dict
+        :param scoped_config: This is the dict representing the configuration
+            for the current profile for the session.
+
+        :type client_name: str
+        :param client_name: client name.
+
+        :type environ: dict
+        :param environ: A mapping to use for environment variables. If this
+            is not provided it will default to use os.environ.
+        """
+        self._full_config = full_config
+        self._scoped_config = scoped_config
+        self._client_name = client_name
+
+        if environ is None:
+            environ = os.environ
+        self._environ = environ
+
+    def provide(self):
+        """Lookup the configured endpoint URL.
+
+        The order is:
+
+        1. The value provided by a service-specific environment variable.
+        2. The value provided by the global endpoint environment variable
+           (AWS_ENDPOINT_URL).
+        3. The value provided by a service-specific parameter from a services
+           definition section in the shared configuration file.
+        4. The value provided by the global parameter from a services
+           definition section in the shared configuration file.
+        """
+        for location in self._ENDPOINT_URL_LOOKUP_ORDER:
+            logger.debug(
+                'Looking for endpoint for %s via: %s',
+                self._client_name,
+                location,
+            )
+
+            endpoint_url = getattr(self, f'_get_endpoint_url_{location}')()
+
+            if endpoint_url:
+                logger.info(
+                    'Found endpoint for %s via: %s.',
+                    self._client_name,
+                    location,
+                )
+                return endpoint_url
+
+        logger.debug('No configured endpoint found.')
+        return None
+
+    def unhyphenize_service_name(self, service_name):
+        return service_name.replace('-', ' ')
+
+    def _get_service_id(self, client_name):
+        client_name = utils.handle_service_name_alias(client_name)
+
+        client_name = (
+            utils._CLIENT_NAME_TO_HYPHENIZED_SERVICE_ID_OVERRIDES.get(
+                client_name, client_name
+            )
+        )
+
+        return self.unhyphenize_service_name(client_name)
+
+    def _snakecase_service_id(self, service_id):
+        return service_id.replace(' ', '_')
```

Comment:
Let's just combine these two methods to be:
```python
def _get_snake_case_service_id(self):
    client_name = utils.handle_service_name_alias(client_name)
    client_name = (
        utils._CLIENT_NAME_TO_HYPHENIZED_SERVICE_ID_OVERRIDES.get(
            self._client_name, self._client_name
        )
    )
    return client_name.replace('-', '_')
```
Within this function, it would then be good to add a comment about how we are supposed to be using the service id but instead deriving it from the client name and what assumptions/properties we are expecting in order to do this.

The main reason that I suggest this is that the logic seems fairly complicated now as there are three or four different methods now in how we are transforming the service id; when in reality, we only just need shared logic for getting the snake cased service id and from there can call upper and lower on it.